### PR TITLE
Arrays support

### PR DESCRIPTION
**Pull request by CookiePLMonster**
Arrays with int/float types now seem to work. The generated code matches SA even for III and VC targets - which in turn (almost) matches miss2 array output. miss2 output missed a data type though, so generated arrays were ambiguous.

Unlike miss2, array indexing is zero-based. 1-based arrays are not available, since there is no code to keep compatibility with.

### CODE REVIEW

**Comment by CookiePLMonster** [comment]
I have no clue how to fix those build errors - VS14 builds those fine :(

File: src/grammar/gta3script.g [hunk 1]
```diff
@@ -195,7 +195,7 @@ expressionStatement
 	:	(id=identifier opa1=assignmentOperators1 a=argument newLine)
 		->  ^($opa1 $id $a)
 	|	(id=identifier opa2=assignmentOperators2 a=argument newLine)
-		->  ^(OP_EQ $id ^($opa2 $id $a))
+		->  ^(OP_EQ $id ^($opa2 ^($id) $a))
```
Comment (thelink2012) [inline src/grammar/gta3script.g]: TODO CHECK

File: src/commands.cpp [hunk 2]
```diff
@@ -253,7 +260,47 @@ void annotate_internal(const Commands& commands, const SymTable& symbols, const
 
             case NodeType::Array:
             {
-                // TODO array SA
+                // Annotate first child as usual, and apply some special annotation to second child
```
Comment (thelink2012) [inline src/commands.cpp]: TODO CHECK

File: src/commands.cpp [hunk 3]
```diff
@@ -173,7 +173,14 @@ const Command& match_internal(const Commands& commands, const SymTable& symbols,
                     bad_alternative = !(argtype_matches(it_alter_arg->type, ArgType::Float) && it_alter_arg->allow_constant);
                     break;
                 case NodeType::Array:
-                    // TODO array SA
+                    try
+                    {
+                        match_identifier( (*it_target_arg)->child(0), commands, *it_alter_arg, symbols, scope_ptr);
```
Comment (thelink2012) [inline src/commands.cpp]: TODO CHECK

File: src/compiler.hpp [hunk 4]
```diff
@@ -482,7 +482,26 @@ struct CompilerContext
 
             case NodeType::Array:
             {
-                // TODO array SA
+                auto opt_var = arg_node.child(0).maybe_annotation<shared_ptr<Var>>();
+                if(opt_var)
+                {
+                    if(auto opt_int = arg_node.child(1).maybe_annotation<int32_t>())
+                    {
+                        return CompiledVar{ std::move(*opt_var), std::move(*opt_int) };
```
Comment (thelink2012) [inline src/compiler.hpp]: TODO GCC/CLANG FIX

File: src/compiler.hpp [hunk 5]
```diff
@@ -482,7 +482,26 @@ struct CompilerContext
...
+                    }
+                    else if(auto opt_index = arg_node.child(1).maybe_annotation<shared_ptr<Var>>())
+                    {
+                        return CompiledVar{ std::move(*opt_var), std::move(*opt_index) };
```
Comment (thelink2012) [inline src/compiler.hpp]: TODO GCC/CLANG FIX

**Comment by thelink2012** [comment]
TODO -farrays

**Pull request merged**